### PR TITLE
fix: voluntary exit signature in post-deneb networks

### PR DESCRIPTION
**Pull request by y0sher**
### Issue:
Following the Deneb fork on the Ethereum Holesky network, the domain in signatures of voluntary exit messages was fixed to the Capella fork and epoch. This change was not accounted for in our codebase, leading to a critical bug where voluntary exits failed on networks post-Deneb.

### Solution:
This PR introduces a fix that aligns our handling of voluntary exit signatures with the updated domain requirements set forth by the Deneb fork. It ensures voluntary exits are correctly processed in networks operating under the post-Deneb specifications.

### CODE REVIEW

File: beacon/goclient/signing.go [hunk 1]
```diff
@@ -11,6 +13,41 @@ import (
 	"github.com/pkg/errors"
 )
 
+func (gc *goClient) computeVoluntaryExitDomain(ctx context.Context,
+) (phase0.Domain, error) {
```
Comment (nkryuchkov) [inline beacon/goclient/signing.go:17]: Why do we need this to be the new line?

File: beacon/goclient/signing.go [hunk 2]
```diff
@@ -11,6 +13,41 @@ import (
...
+	forkData := &phase0.ForkData{
+		CurrentVersion: forkVersion,
+	}
+
+	response, err := gc.client.Genesis(ctx, &api.GenesisOpts{})
+	if err != nil {
+		return phase0.Domain{}, errors.Wrap(err, "failed to obtain genesis")
+	}
+
+	forkData.GenesisValidatorsRoot = response.Data.GenesisValidatorsRoot
```
Comment (nkryuchkov) [inline beacon/goclient/signing.go:37]: It likely won't ever happen, but I'd also make sure that `response.Data != nil`

**Review by nkryuchkov — APPROVED** [review]

**Review by nkryuchkov — APPROVED** [review]

**Review by zippoxer — APPROVED** [review]

File: beacon/goclient/signing.go [hunk 3]
```diff
@@ -11,6 +13,45 @@ import (
 	"github.com/pkg/errors"
 )
 
+func (gc *goClient) computeVoluntaryExitDomain(ctx context.Context) (phase0.Domain, error) {
+	specs, err := gc.client.Spec(gc.ctx, &api.SpecOpts{})
+	if err != nil {
+		return phase0.Domain{}, errors.Wrap(err, "failed to obtain fork schedule")
+	}
```
Comment (olegshmuelov) [inline beacon/goclient/signing.go:20]: we should check for spec response is nil

File: beacon/goclient/signing.go [hunk 4]
```diff
@@ -11,6 +13,45 @@ import (
...
+
+	forkData := &phase0.ForkData{
+		CurrentVersion: forkVersion,
+	}
+
+	response, err := gc.client.Genesis(ctx, &api.GenesisOpts{})
+	if err != nil {
+		return phase0.Domain{}, errors.Wrap(err, "failed to obtain genesis")
+	}
```
Comment (olegshmuelov) [inline beacon/goclient/signing.go:37]: same here

**Review by olegshmuelov — CHANGES_REQUESTED** [review]

File: beacon/goclient/signing.go (hunk 3 above)
Comment (olegshmuelov) [reply beacon/goclient/signing.go:20]: specs.Data as well

File: beacon/goclient/signing.go [hunk 5]
```diff
@@ -11,6 +13,45 @@ import (
...
+		return phase0.Domain{}, errors.Wrap(err, "failed to obtain fork schedule")
+	}
+
+	// TODO: consider storing fork version and genesis validators root in goClient
+	//		instead of fetching it every time
+
+	forkVersion, ok := specs.Data["CAPELLA_FORK_VERSION"].(phase0.Version)
+	if !ok {
+		return phase0.Domain{}, errors.New("failed to decode fork version")
+	}
```
Comment (olegshmuelov) [inline beacon/goclient/signing.go:28]:
```suggestion
tmp, exists := specs.Data["CAPELLA_FORK_VERSION"]
	if !exists {
		return phase0.Domain{}, fmt.Errorf("capella fork version not known by chain")
	}

	forkVersion, isForkVersion := tmp.(phase0.Version)
	if !isForkVersion {
		return phase0.Domain{}, fmt.Errorf("could not obtain CAPELLA_FORK_VERSION")
	}
```

File: beacon/goclient/signing.go [hunk 6]
```diff
@@ -11,6 +13,45 @@ import (
 	"github.com/pkg/errors"
 )
 
+func (gc *goClient) computeVoluntaryExitDomain(ctx context.Context) (phase0.Domain, error) {
+	specs, err := gc.client.Spec(gc.ctx, &api.SpecOpts{})
+	if err != nil {
+		return phase0.Domain{}, errors.Wrap(err, "failed to obtain fork schedule")
```
Comment (olegshmuelov) [inline beacon/goclient/signing.go:19]: IIRC we decide to not use errors package, but use fmt

File: beacon/goclient/signing.go [hunk 7]
```diff
@@ -11,6 +13,45 @@ import (
...
+		return phase0.Domain{}, errors.New("failed to decode fork version")
+	}
+
+	forkData := &phase0.ForkData{
+		CurrentVersion: forkVersion,
+	}
+
+	response, err := gc.client.Genesis(ctx, &api.GenesisOpts{})
+	if err != nil {
+		return phase0.Domain{}, errors.Wrap(err, "failed to obtain genesis")
```
Comment (olegshmuelov) [inline beacon/goclient/signing.go:36]: IIRC we decide to not use errors package, but use fmt

File: beacon/goclient/signing.go [hunk 8]
```diff
@@ -11,6 +13,45 @@ import (
...
+	}
+	if response.Data == nil {
+		return phase0.Domain{}, errors.New("failed to obtain genesis")
```
Comment (olegshmuelov) [inline beacon/goclient/signing.go:39]: IIRC we decide to not use errors package, but use fmt

File: beacon/goclient/signing.go [hunk 9]
```diff
@@ -11,6 +13,45 @@ import (
...
+	}
+	forkData.GenesisValidatorsRoot = response.Data.GenesisValidatorsRoot
+
+	root, err := forkData.HashTreeRoot()
+	if err != nil {
+		return phase0.Domain{}, errors.Wrap(err, "failed to calculate signature domain")
```
Comment (olegshmuelov) [inline beacon/goclient/signing.go:45]: IIRC we decide to not use errors package, but use fmt

**Review by moshe-blox — DISMISSED** [review]

File: beacon/goclient/signing.go [hunk 10]
```diff
@@ -11,6 +14,58 @@ import (
 	"github.com/pkg/errors"
 )
 
+func (gc *goClient) computeVoluntaryExitDomain(ctx context.Context) (phase0.Domain, error) {
+	specs, err := gc.client.Spec(gc.ctx, &api.SpecOpts{})
+	if err != nil {
+		return phase0.Domain{}, errors.Wrap(err, "failed to obtain specs")
```
Comment (nkryuchkov) [inline beacon/goclient/signing.go:20]: If we use `fmt.Errorf`, I think we should use it in all cases

File: beacon/goclient/signing.go [hunk 11]
```diff
@@ -11,6 +14,58 @@ import (
...
+	}
+	if specs == nil {
+		return phase0.Domain{}, fmt.Errorf("failed to obtain specs")
+	}
+	if specs.Data == nil {
+		return phase0.Domain{}, fmt.Errorf("failed to obtain specs")
```
Comment (nkryuchkov) [inline beacon/goclient/signing.go:26]: Perhaps a bit different text would help to understand what exactly is nil if this error appears in logs

File: beacon/goclient/signing.go [hunk 12]
```diff
@@ -11,6 +14,58 @@ import (
...
+		return phase0.Domain{}, fmt.Errorf("failed to decode capella fork version")
+	}
+
+	forkData := &phase0.ForkData{
+		CurrentVersion: forkVersion,
+	}
+
+	response, err := gc.client.Genesis(ctx, &api.GenesisOpts{})
+	if err != nil {
+		return phase0.Domain{}, errors.Wrap(err, "failed to obtain genesis")
```
Comment (nkryuchkov) [inline beacon/goclient/signing.go:47]: fmt.Errorf

File: beacon/goclient/signing.go [hunk 13]
```diff
@@ -11,6 +14,58 @@ import (
...
+		return phase0.Domain{}, fmt.Errorf("failed to obtain genesis")
+	}
+	if response.Data == nil {
+		return phase0.Domain{}, fmt.Errorf("failed to obtain genesis")
+	}
+	forkData.GenesisValidatorsRoot = response.Data.GenesisValidatorsRoot
+
+	root, err := forkData.HashTreeRoot()
+	if err != nil {
+		return phase0.Domain{}, errors.Wrap(err, "failed to calculate signature domain")
```
Comment (nkryuchkov) [inline beacon/goclient/signing.go:59]: fmt

File: beacon/goclient/signing.go [hunk 14]
```diff
@@ -11,6 +14,58 @@ import (
...
+	}
+
+	var domain phase0.Domain
+	copy(domain[:], spectypes.DomainVoluntaryExit[:])
+	copy(domain[4:], root[:])
```
Comment (nkryuchkov) [inline beacon/goclient/signing.go:64]: nit: I don't think we'll ever change domain length but perhaps we could make its length abstract, but I don't mind leaving it as it is
```suggestion
	copy(domain[len(spectypes.DomainVoluntaryExit):], root[:])
```

File: beacon/goclient/signing.go [hunk 15]
```diff
@@ -11,6 +14,58 @@ import (
...
+
+	response, err := gc.client.Genesis(ctx, &api.GenesisOpts{})
+	if err != nil {
+		return phase0.Domain{}, errors.Wrap(err, "failed to obtain genesis")
+	}
+	if response == nil {
+		return phase0.Domain{}, fmt.Errorf("failed to obtain genesis")
+	}
+	if response.Data == nil {
+		return phase0.Domain{}, fmt.Errorf("failed to obtain genesis")
```
Comment (nkryuchkov) [inline beacon/goclient/signing.go:53]: I'd consider changing the text

File: beacon/goclient/signing.go [hunk 16]
```diff
@@ -1,7 +1,10 @@
 package goclient
 
 import (
+	"context"
 	"crypto/sha256"
+	"fmt"
+	"github.com/attestantio/go-eth2-client/api"
```
Comment (nkryuchkov) [inline beacon/goclient/signing.go:7]: nit: I think this should be near "github.com/attestantio/go-eth2-client/spec/phase0" to be able to quickly see the scope of all imports from the library

**Review by nkryuchkov — DISMISSED** [review]
LGTM. No major requests, just minor nitpicks, hence approving

File: beacon/goclient/signing.go (hunk 14 above)
Comment (nkryuchkov) [reply beacon/goclient/signing.go:64]: Also, I just want to confirm something:

`phase0.Domain`' length is 32, spectypes.DomainVoluntaryExit is 4 and root is 32, so last 4 bytes of `root` won't be copied, but as I understand it's ok?

**Review by olegshmuelov — APPROVED** [review]

File: beacon/goclient/signing.go [hunk 17]
```diff
@@ -1,16 +1,71 @@
...
+		return phase0.Domain{}, fmt.Errorf("genesis response is nil")
+	}
+	if genesisResponse.Data == nil {
+		return phase0.Domain{}, fmt.Errorf("genesis response data is nil")
+	}
+	forkData.GenesisValidatorsRoot = genesisResponse.Data.GenesisValidatorsRoot
+
+	root, err := forkData.HashTreeRoot()
+	if err != nil {
+		return phase0.Domain{}, errors.Wrap(err, "failed to calculate signature domain")
```
Comment (nkryuchkov) [inline beacon/goclient/signing.go:59]: Still using errors

Comment (nkryuchkov) [reply beacon/goclient/signing.go:59]: @olegshmuelov

**Review by nkryuchkov — APPROVED** [review]

**Review by olegshmuelov — APPROVED** [review]

File: beacon/goclient/signing.go (hunk 17 above)
Comment (y0sher) [reply beacon/goclient/signing.go]: I fixed it

**Review by nkryuchkov — APPROVED** [review]

**Pull request merged**